Added  kong-plugin-set-dynamic-target-host plugin documentation (#3366)
Kong/docs.konghq.com: app/_hub/flash/set-dynamic-upstream-host/index.md

---
name: Set Dynamic Upstream Host

> publisher: Flash

Reviewer: Where does the name `Flash` come from? I'd have expected `Anup Krai` in this field based on the GitHub org

Author: This plugin is developed by two contributors and team name is `Flash`. So instead of giving individual team member name we gave the team name here. Please let us know if any concerns in this.

version: 1.0.0

desc: Constructs the upstream hostname dynamically based on the incoming request parameters
description: |
  This plugin can be used to dynamically construct upstream hostname and port number based on the key identifier passed in the incoming request. If the same upstream api is deployed in different servers or data centers, then this plugin can form the hostname for the upstream api dynamically to route it to particular server or data center without making any change in Kong route or service.
  This plugin extracts the key identifier from the incoming request and following are the supported parameters -
  * Header param
  * Query param
  * Path param
  * Body param
  
  
support_url: https://github.com/anup-krai/kong-plugin-set-dynamic-target-host/issues
source_url: https://github.com/anup-krai/kong-plugin-set-dynamic-target-host
license_type: Apache-2.0 

type: plugin
categories:
  - traffic-control

> kong_version_compatibility:

Reviewer: Has the plugin been tested with all of these versions? If not, we should limit this compatibility matrix only to those versions tested

Author: Updated the tested versions for this plugin.

    community_edition:
      compatible:
        - 2.4.x
        - 2.3.x
        - 2.2.x
        - 2.1.x
        - 2.0.x
        - 1.5.x
        - 1.4.x
        - 1.3.x
        - 1.2.x
        - 1.1.x
        - 1.0.x
    enterprise_edition:
      compatible:
        - 2.4.x
        - 2.3.x
        - 2.2.x
        - 2.1.x
        - 1.5.x
        - 1.3-x

params:
  name: set-target-host
  service_id: true
  route_id: false
  consumer_id: false
  protocols: ["http", "https"]
  dbless_compatible: true
    
  config:
    - name: upstream_host
      required: true
      default:

>       value_in_examples: "nodenumber.org"

Reviewer: Can we use `example.com` in any domain examples? https://datatracker.ietf.org/doc/html/rfc2606#section-3

Author: Updated the domain to example.org in all the domain examples.

        Based on the path index, `httpbin` string will be extracted from the uri.
        Final upstream hostname formed here is `httpbin.org` and Kong will make a call to this host.
    - name: body_param
      required: semi
      value_in_examples:
      default:
      datatype: string
      description: |
        Request body parameter name which will be used to form the upstream host.
        Only `application/json` and `application/x-www-form-urlencoded` content types are supported. For Json message, the field name or Json path needs to be passed. 
        For eg: If Json message is `{"target": "httpbin"}`
        and `config.body_param=target` then `httpbin` value from the `target` field in the json will be used to form the upstream host. 
        In case of duplicates (target field in json) the entire Json path needs to be provided. 
        For form-urlencoded payload if `--data-urlencode 'target=httpbin'` and `config.body_param=target` then `httpbin` value from the `target` field in the form body will be used to construct the upstream host.
  extra: |

>     Note : At any point of time only one parameter is supported and at least one is required.

Reviewer: ```suggestion
    Note : Only one of `header`, `query_arg`, `path_index` or `body_param` can be provided at one time, and at least one is required.
```
